eta/core/serial.py

```diff
@@ -812,6 +818,550 @@ def _validate_dict(cls, d):
         return etau.get_class(d["_CLS"])
 
 
+class BigMixin(object):
+    '''Mixin class for "big" Serializable classes, which store their elements
+    on disk rather than in-memory.
+
+    Subclasses must call BigMixin's constructor to properly initialize the
+    Big object.
+
+    `backing_dir` is an optional keyword argument for any methods that require
+    initializing a new instance of a Big iterable. When `backing_dir` is
+    `None`, a temporary directory is used and is deleted when the Big iterable
+    is garbage collected.
+
+    `move()` can be used to move the Big iterable between a set (persistent)
+    backing directory and a temporary backing directory.
+    '''
+
+    def __init__(self, backing_dir):
+        '''Initializes the base BigMixin.
+
+        Args:
+            backing_dir: the backing directory to use, or None if using
+                temporary storage
+        '''
+        self._backing_dir = None
+        self._uses_temporary_storage = False
+        self._set_backing_dir(backing_dir)
+
+    def __del__(self):
+        if self.uses_temporary_storage and os.path.exists(self.backing_dir):
+            etau.delete_dir(self.backing_dir)
+
+    @property
+    def backing_dir(self):
+        '''The backing directory for this Big iterable.'''
+        return self._backing_dir
+
+    @property
+    def uses_temporary_storage(self):
+        '''Whether this Big iterable is backed by temporary storage.'''
+        return self._uses_temporary_storage
+
+    def clear(self):
+        '''Deletes all elements from the Big iterable.'''
+        super(BigSet, self).clear()
+        etau.delete_dir(self.backing_dir)
+        etau.ensure_dir(self.backing_dir)
+
+    def copy(self, backing_dir=None):
+        '''Creates a deep copy of this Big iterable backed by the given
+        directory.
+
+        Args:
+            backing_dir: optional backing directory to use for the new big
+                iterable. If provided, it must be empty or non-existent
+
+        Returns:
+            a Big iterable
+        '''
+        new_big = self.empty(backing_dir=backing_dir)
+        new_big.add_iterable(self)
```

Comment:
For clarity, `BigMixin` should declare `add_iterable()` as an abstract method that subclasses need to implement

Reply:
Done!